pkg/controller/node/nodecontroller.go

```diff
@@ -555,6 +565,88 @@ func (nc *NodeController) Run(stopCh <-chan struct{}) {
 	<-stopCh
 }
 
+// addOrRemoveNoScheduleTaintByCondition taint or remove NoSchedule taint on this node based on it's condition change.
+// return:
+// 1. if node changed to schedulable
+// 2. if node changed to unschedulable
+func (nc *NodeController) addOrRemoveNoScheduleTaintByCondition(node *v1.Node) (bool, bool) {
+	var changedToUnSchedulable, changedToSchedulable bool
+	observedMap := map[v1.NodeConditionType]*v1.NodeCondition{}
+	_, observedMap[v1.NodeReady] = nodeutil.GetNodeCondition(&node.Status, v1.NodeReady)
+	_, observedMap[v1.NodeOutOfDisk] = nodeutil.GetNodeCondition(&node.Status, v1.NodeOutOfDisk)
+	_, observedMap[v1.NodeNetworkUnavailable] = nodeutil.GetNodeCondition(&node.Status, v1.NodeNetworkUnavailable)
+
+	// If there are saved node status, it only make sense to check conditions when some change happened.
+	if savedNodeStatus, found := nc.nodeStatusMap[node.GetName()]; found {
```

Comment:
This doesn't work. The Node you're getting here is the same Node that was used to set Status in `nodeStatusMap`. It means that there will never be any update here. When fixing this please add regression test for this case.

I think the only thing you can compare would be a set of Taints that are present on the Node.